### PR TITLE
Katharopoulos2020lineartransformer Srinath

### PR DESCRIPTION
**Pull request by sritank**
This is my first draft. Please review.

### CODE REVIEW

File: _summaries/katharopoulos2020lineartransformer_2.md [hunk 1]
```diff
@@ -0,0 +1,58 @@
+---
+layout: summary
+title: Summary
+paper: {{katharopoulos2020lineartransformer}}
+author:  sritank
+score:  10
+---
+
+## What is the core idea?
+A new linear transformer model is introduced that uses a kernel based self-attention method that reduces the memory footprint of the model. They also linearize the masking of the attention computation making it run in linear time.
```
Comment (saikm200022) [inline _summaries/katharopoulos2020lineartransformer_2.md:10]: Another important idea in the paper was using their causal masked attention to express a transformer layer as a Recurrent Neural Network. Might be a good idea to add that to the core idea section.

File: _summaries/katharopoulos2020lineartransformer_2.md [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+![softmax kernel](./katharopoulos2020lineartransformer_2b.png)
+
+Instead of a softmax similarity function, the authors propose using a kernel function of a feature representation \phi(x). So the new self attention function is:
+
+![linear self attention function](./katharopoulos2020lineartransformer_2c.png)
+
+The above attention function has computation cost of O(N) where N is the sequence length. It's memory footprint is also lower as the terms can be reused for every query hence we don't need to store the full attention matrix making it's memory complexity O(N). The authors state that for small sequences the below feature map, phi(x), was used and shown to perform on par to the full traditional transformer.
+
+They further change the masking function to span until the i-th position of the query instead of the whole sequence (i.e. all N inputs). They called this ``causal'' masking. They further show how their formulation of a transformer with causal masking can be understood as an RNN.
```
Comment (saikm200022) [inline _summaries/katharopoulos2020lineartransformer_2.md:28]: I think you should mention the motivation behind why they even want to represent the transformer layer as an RNN. Something like - the linear transformer models allows parallelization during training, but also allows much faster inference.

File: _summaries/katharopoulos2020lineartransformer_2.md [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+![causal masking](./katharopoulos2020lineartransformer_2d.png)
+
+
```
Comment (saikm200022) [inline _summaries/katharopoulos2020lineartransformer_2.md:32]: Just for some more context, you could also specify which part of the self-attention computation is being stored as a hidden state in the RNN representation of the transformer layer - phi (K_j) V^T_j.

File: _summaries/katharopoulos2020lineartransformer_2.md [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+- symbol copying task
+
+The authors found that their transformer had a lower computational time (log linear) for a given sequence length compared to other attention functions in literature, including softmax. The memory footprint of their linear transformer was also significantly lower than other in literature, including softmax.
+
+![copy task performance](./katharopoulos2020lineartransformer_2e.png)
+
+- Image generation
+
+They tested their transformer on the MNIST and CIFAR-10 dataset. They were able to achieve performance on par with softmax attention in terms of bits per dimension and perplexity. There network excelled at speed, being able to generate images around 300 times faster with constant memory per image on MNIST and ~4000 times faster on CIFAR-10.
```
Comment (saikm200022) [inline _summaries/katharopoulos2020lineartransformer_2.md:45]: Like the other experimental result sections, it might be a good idea to go ahead and add table 2 from the paper since you are referencing its results.

File: _summaries/katharopoulos2020lineartransformer_2.md [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+- Automatic speech recognition
+
+They trained the linear transformer with connectionist temporal classification (CTC) loss on 80hrs of the WSJ dataset. They find that the Linear transformer converges to a Prediction Error rate (PER) that is lower than LSH-4 and Bi-LSTM. However, their Linear transformer has a PER that is worse than softmax attention. They however beat all other models in terms of training time and evaluation time.
+
+![ASR performance](./katharopoulos2020lineartransformer_2f.png)
+
+
+## TL;DR
+- Introduce a linear transformer with a different attention function with causal masking
+- Linear transformer has much lower memory footprint and computational cost and is faster to train and evaluate than softmax attention transformer
```
Comment (saikm200022) [inline _summaries/katharopoulos2020lineartransformer_2.md:56]: One thing you could emphasize in this section (which is mentioned a few times in the paper) is that their "Linear Transformer" performed just as good as the traditional softmax transformer despite the significant speed up.

File: _summaries/katharopoulos2020lineartransformer_2.md [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+
+## How is it realized (technically)?
+
+They change the attention mechanism from the traditional softmax function to a dot product feature map. In a traditional transformer, given the queries Q, keys K and values V, the self attention function A_l can be calculated as:
+
+![self attention function](./katharopoulos2020lineartransformer_2a.png)
+
+where,
+
+![softmax kernel](./katharopoulos2020lineartransformer_2b.png)
```
Comment (saikm200022) [inline _summaries/katharopoulos2020lineartransformer_2.md:20]: I would mention something here about how the paper used the associativity property of matrices to be able to simplify to this equation where the summation results involving the K and V matrices are reused for every query.

File: _summaries/zeiler2012adadelta_1.md [hunk 7]
```diff
@@ -1,14 +1,43 @@
 ---
 layout: summary
 title: Summary
-paper: zeiler2012adadelta
+
+paper: {{zeiler2012adadelta}}
```
Comment (saikm200022) [inline _summaries/zeiler2012adadelta_1.md:5]: You don't need the brackets around the paper tag.

**Review by saikm200022 — CHANGES_REQUESTED** [review]
Great summary! Here are my suggestions:

File: _summaries/katharopoulos2020lineartransformer_2.md (hunk 6 above)
Comment (sritank) [reply _summaries/katharopoulos2020lineartransformer_2.md:20]: I did, on line 26. Made it more explicit now

File: _summaries/katharopoulos2020lineartransformer_2.md (hunk 2 above)
Comment (sritank) [reply _summaries/katharopoulos2020lineartransformer_2.md:28]: Not sure I understand. I thought transformers were already parallelizable during training. RNNs and LSTM training are not parallelizable because of their sequential nature which was one of their biggest limitations. I guess their motivations for expressing it as RNN is to better understand how a transformer works and what attention does.

Comment (saikm200022) [reply _summaries/katharopoulos2020lineartransformer_2.md]: I was referring to how in the paper they mentioned how their idea of a "Linear Transformer" had the "best of both worlds" because it allows for parallelization during training and much faster inference compared to standard transformers. 

You can ignore the first sentence of my comment - I think I got some stuff mixed up.

Comment (sritank) [reply _summaries/katharopoulos2020lineartransformer_2.md]: Got it. I will mention this during the presentation

**Review by saikm200022 — APPROVED** [review]
Looks good!

**Comment by zhaoyue-zephyrus** [comment]
@sritank  Could you remove all files related to adadelta?

**Comment by sritank** [comment]
@zhaoyue-zephyrus I've removed the images

**Review by zhaoyue-zephyrus — APPROVED** [review]

**Comment by zhaoyue-zephyrus** [comment]
@sritank  LGTM. Thank you!
